titus-client/src/main/java/com/netflix/titus/runtime/connector/jobmanager/PCollectionJobSnapshot.java

```diff
@@ -0,0 +1,332 @@
+
+    private static final JobSnapshot EMPTY = newInstance("empty", HashTreePMap.empty(), Collections.emptyMap());
+
+    private final PMap<String, Job<?>> jobsById;
+    private final PMap<String, PSequence<Task>> tasksByJobId;
+    private final PMap<String, Task> taskById;
+
+    private final String signature;
+
+    // Deprecated values computed lazily
+    private volatile List<Job<?>> allJobs;
+    private final Lock allJobsLock = new ReentrantLock();
+
+    private volatile List<Task> allTasks;
+    private final Lock allTasksLock = new ReentrantLock();
+
+    private volatile List<Pair<Job<?>, List<Task>>> allJobTaskPairs;
+    private final Lock allJobTaskPairsLock = new ReentrantLock();
+
+    public static JobSnapshot empty() {
+        return EMPTY;
+    }
+
+    public static JobSnapshot newInstance(String snapshotId, Map<String, Job<?>> jobsById, Map<String, List<Task>> tasksByJobId) {
+        Map<String, PSequence<Task>> mTasksByJobId = new HashMap<>();
+        Map<String, Task> taskById = new HashMap<>();
+        tasksByJobId.forEach((jobId, tasks) -> {
+            TreePVector<Task> pTasks = TreePVector.from(tasks);
+            mTasksByJobId.put(jobId, pTasks);
+            tasks.forEach(task -> taskById.put(task.getId(), task));
+        });
+
+        return new PCollectionJobSnapshot(
+                snapshotId,
+                HashTreePMap.from(jobsById),
+                HashTreePMap.from(mTasksByJobId),
+                HashTreePMap.from(taskById)
+        );
+    }
+
+    private PCollectionJobSnapshot(String snapshotId, PMap<String, Job<?>> jobsById, PMap<String, PSequence<Task>> tasksByJobId,
+                                   PMap<String, Task> taskById) {
+        super(snapshotId);
+        this.jobsById = jobsById;
+        this.tasksByJobId = tasksByJobId;
+        this.taskById = taskById;
+        this.signature = computeSignature();
+    }
+
+    public Map<String, Job<?>> getJobMap() {
+        return jobsById;
+    }
+
+    public List<Job<?>> getJobs() {
+        if (allJobs != null) {
+            return allJobs;
+        }
+        allJobsLock.lock();
+        try {
+            List<Job<?>> acc = new ArrayList<>();
+            jobsById.forEach((jobId, job) -> acc.add(job));
+            this.allJobs = Collections.unmodifiableList(acc);
+        } finally {
+            allJobsLock.unlock();
+        }
+        return allJobs;
+    }
+
+    public Optional<Job<?>> findJob(String jobId) {
+        return Optional.ofNullable(jobsById.get(jobId));
+    }
+
+    @Override
+    public Map<String, Task> getTaskMap() {
+        return taskById;
+    }
+
+    public List<Task> getTasks() {
+        if (allTasks != null) {
+            return allTasks;
+        }
+        allTasksLock.lock();
+        try {
+            List<Task> acc = new ArrayList<>();
+            taskById.forEach((taskId, task) -> acc.add(task));
+            this.allTasks = acc;
```

Comment:
minor: `minor: `this.allTasks = Collections.unmodifiableList(new ArrayList<>(taskById.values()));`